### ISSUE TITLE
setZoomLevel causes a webview to no longer fill the full page

### ISSUE THREAD

**Issue by bpasero** [electron/electron#7375]
- Electron version: 1.3.7
- Operating system: Mac OS

When you set the zoom level on the webframe that has a webview element, what the webview is showing no longer spans the full width of the page:

Without zoom:
![screen shot 2016-09-28 at 07 36 43](https://cloud.githubusercontent.com/assets/900690/18901811/6e024502-854e-11e6-98b6-b68e7deaae3e.png)

With zoom:
![image](https://cloud.githubusercontent.com/assets/900690/18901805/5b344128-854e-11e6-91b3-ae95e57c3d78.png)

HTML:

``` html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Hello World!</title>
    <script src="./index.js"></script>
  </head>
  <body>
    <webview src="http://www.github.com" id="foo" />
  </body>
</html>
```

JS:

``` javascript
setTimeout(function () {
    var webFrame = require('electron').webFrame;
    webFrame.setZoomLevel(1);
}, 2000);
```

**Comment by andymartinwork** [comment]
Might potentially be related to #7458 and #6958.

**Comment by alvarezn** [comment]
I ran into this as well, there seem to be a quite a few bugs related to the sizing of webviews when zoom factor/level are applied, all likely related.  The parent page seems to size the element correctly, but the content loaded by the webview is given an incorrect size to render in.  If its supposed to have 10% more space it seems to have 10% less instead - with the opposite true as well.  This seems to have been introduced in v1.2.4.  v1.2.3 correctly sizes the scaled webviews.

Very apparent when you scale the page and give the webview a border with content source which scrolls in any (or both) directions.  Scrollbars will be nowhere near the border.

**Comment by zcbenz** [comment]
This is an upstream bug, we should make a reproducible example with Chrome extension and report this to [Chromium's issues](https://crbug.com)/

**Comment by ccnokes** [comment]
I was about to open the same issue 😸 . I'm seeing it in 1.4.5 on macOS.

For anyone looking for a workaround, I've found that you can set the zoom on just the webview (e.g. `webview.setZoomFactor(1.1)`) but leave the zoom untouched in the parent renderer. If you have additional UI in the parent renderer and you want that zoomed as well, you can always use the CSS `zoom` property. Setting `zoom` on the body doesn't seem to affect the webview.

**Comment by bpasero** [comment]
With the new Electron 1.4.7 release and [Allow webview guests to be resized manually](https://github.com/electron/electron/pull/7658) landing, I wonder if this can be worked around from the application. I did not have much luck though from my initial testing.

**Comment by martinpichaat4mation** [comment]
Hi there @bpasero ,

I have actually experienced the same issue and spent quite a while figuring this out. I'm not sure if you have by now, but just for a reference or anybody stumbling on this issue as well, the trick is to not set the height and width using percentage (or calc(_ ...) as I was using)_ but to use the vh and vw units.

`<webview style="height: 100vh; width: 100vw" ...`

Additionally, your html, body elements should be:

`
body, html { height: 100%; width: 100%; min-height: 100%; min-width: 100% }
`

**Comment by bpasero** [comment]
@martinpichaat4mation meanwhile we have "solved" this by updating the webview size dynamically when the outer container changes in size. 

@mjbvz fyi on https://github.com/electron/electron/issues/7375#issuecomment-306663559

**Comment by boljen** [comment]
Here's a React component which works for me (note: my app uses setZoomFactor for zooming, not setZoomLevel);

https://gist.github.com/boljen/ce879f0a8260b67bf42b65554673115b

**Comment by craxal** [comment]
@martinpichaat4mation Your suggestion only works when the webview is meant to fill the entire window. It doesn't work for webviews that need to share window space with other elements.

@bpasero Do you have a code sample you could share?

**Comment by bpasero** [comment]
@craxal see [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/html/browser/webview.ts#L364).

**Comment by craxal** [comment]
@bpasero Seems to work in some basic scenarios, but it doesn't work well with things like collapsing panels get involved. The webview briefly resizes correctly then snaps back to the incorrect dimensions.

**Comment by akshayhashedIn** [comment]
> @craxal see [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/html/browser/webview.ts#L364).

@bpasero The link seems to be broken.

**Comment by craxal** [comment]
@zcbenz Has this bug been fixed yet? I've been experimenting with v4.2.5, and I don't think I see this issue come up anymore.

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

If this is still reproducible on a supported version, please open a new issue with any other new information that a maintainer should know.

Thank you for taking the time to report this issue and helping to make Electron better! Your help is appreciated.
